zero/src/block_interval.rs

```diff
@@ -1,16 +1,48 @@
 use std::pin::Pin;
 use std::sync::Arc;
+use std::{future::Future, ops::Range};
 
-use alloy::primitives::B256;
-use alloy::rpc::types::eth::BlockId;
-use alloy::{hex, providers::Provider, transports::Transport};
+use alloy::rpc::types::{eth::BlockId, Block};
 use anyhow::{anyhow, Result};
 use async_stream::try_stream;
 use futures::Stream;
+#[cfg(test)]
+use mockall::automock;
 use tracing::info;
 
-use crate::parsing;
-use crate::provider::CachedProvider;
+#[cfg_attr(test, automock)]
+pub trait BlockIntervalProvider<T> {
```

Comment:
Not sure I understand. Why do you need `<T>` if it is never used in trait definition or `block_interval_provider_impl`? Is it just to map it one to one with `Provider<T>`,  equivalent to structs `std::marker::PhantomData<T>`?

Reply:
For `CachedProvider<P,T>` to impl `alloy::providers::Provider`:
```
impl<T, P> Provider<T> for CachedProvider<P, T>
where
    T: Transport + Clone,
    P: Provider<T>,
{
    fn root(&self) -> &RootProvider<T> {
        self.provider.root()
    }
}
```
Hence `P` must constrain `T` here:
```
   impl<P: Provider + Clone> BlockIntervalProvider for P {
```
else:
```
the trait `alloy::providers::Provider` is not implemented for `CachedProvider<RootProvider<RetryService<Http<Client>>>
```
LMK if you think of a way to avoid the `T` despite the relationship between those 2 generic trait impls and `CachedProvider<P,T>`.